### PR TITLE
tensorflow_backend.py rnn does not work if the output.dtype is different from the input dtype (Embedding)

### PR DESCRIPTION
**Pull request by jarfo**
tensorflow_backend.py rnn does not work if the output.dtype is different from the input.dtype.
For example TimeDistributed(Embeddings) (with fixed batch size)

Example that was not working with the tensorflow backend:
        chars = Input(batch_shape=(opt.batch_size, opt.seq_length, opt.max_word_l), dtype='int32', name='chars')
        chars_embedding = TimeDistributed(Embedding(opt.char_vocab_size, opt.char_vec_size, name='chars_embedding'))(chars)

### CODE REVIEW

**Comment by fchollet** [comment]
Can you add a unit test to demonstrate the issue and show that your changes fix it?

**Comment by jarfo** [comment]
Added a 	TimeDistributed(Embedding) test unit with and without fixed batch size

**Comment by fchollet** [comment]
I think we should actually fix the fact that the TF RNN wouldn't work for ndim > 3; what error are you getting?

**Comment by jarfo** [comment]
Yes. You are right. In fact, in the current Keras version, the problem has nothing to do with the output dim but with the 'int32' dtype of the input in the Embedding layer. The problem is in line 2104 of tensorflow_backend.py that assumes output dtype to be equal to the input dtype. (This is not the case in the Embedding layer)

        output_ta = tensor_array_ops.TensorArray(
            dtype=inputs.dtype,
            size=time_steps,
            tensor_array_name='output_ta')

One solution is to call the step function to get the outputs.dtype:

        outputs, _ = step_function(inputs[0], initial_states + constants)
        output_ta = tensor_array_ops.TensorArray(
            dtype=outputs.dtype,
            size=time_steps,
            tensor_array_name='output_ta')

**Comment by jarfo** [comment]
I am closing the pull the request. Let me know if you want me to open another one with the new proposed solution to the tensorflor_backend rnn output_ta.dtype problem.
